File: internal/exercism/status.go
package exercism

import "github.com/charmbracelet/lipgloss"

// Status is the derived state of an exercise.
type Status int

const (
	NotStarted Status = iota
	InProgress
	Completed
	Published
	Locked
)

// DeriveStatus maps an exercise + optional solution to a Status. The v2 schema
// is unofficial, so unknown non-empty solution statuses fall back to InProgress.
func DeriveStatus(isUnlocked bool, solutionStatus string, hasSolution bool) Status {
	if !hasSolution {
		if isUnlocked {
			return NotStarted
		}
		return Locked
	}
	switch solutionStatus {
	case "completed":
		return Completed
	case "published":
		return Published
	case "started", "iterated":
		return InProgress
	default:
		return InProgress
	}
}

type badge struct {
	glyph string
	color lipgloss.Color
	label string
}

var badges = map[Status]badge{
	NotStarted: {"●", lipgloss.Color("8"), "not started"},
	InProgress: {"◐", lipgloss.Color("3"), "in progress"},
	Completed:  {"✓", lipgloss.Color("2"), "completed"},
	Published:  {"★", lipgloss.Color("6"), "published"},
	Locked:     {"🔒", lipgloss.Color("1"), "locked"},
}

// Badge returns the colored glyph for a status.
func (s Status) Badge() string {
	b := badges[s]
	return lipgloss.NewStyle().Foreground(b.color).Render(b.glyph)
}

// Label returns the human-readable name for a status.
func (s Status) Label() string {
	return badges[s].label
}

// Legend renders the status legend line for list footers.
func Legend() string {
	order := []Status{NotStarted, InProgress, Completed, Published, Locked}
	out := ""
	for i, s := range order {
		if i > 0 {
			out += "   "
		}
		out += s.Badge() + " " + s.Label()
	}
	return out
}

File: lib/xrc/status.ex
defmodule Xrc.Status do
  @moduledoc """
  Derives a single exercise status from the unofficial v2 API shape: an
  `exercise` map (with `is_unlocked`) and an optional `solution` map (with a
  `status` string). The schema is unofficial, so parsing is deliberately lenient.
  """

  @type t :: :not_started | :in_progress | :completed | :published | :locked

  @doc """
  Map an exercise + optional solution to a status atom.

    * no solution and locked -> :locked
    * no solution and unlocked -> :not_started
    * solution "started"/"iterated" -> :in_progress
    * solution "completed" -> :completed
    * solution "published" -> :published
    * any other non-nil solution status -> :in_progress (lenient fallback)
  """
  @spec derive(map(), map() | nil) :: t()
  def derive(exercise, solution)

  def derive(exercise, nil) do
    if Map.get(exercise, "is_unlocked", true), do: :not_started, else: :locked
  end

  def derive(_exercise, solution) when is_map(solution) do
    case Map.get(solution, "status") do
      "completed" -> :completed
      "published" -> :published
      "started" -> :in_progress
      "iterated" -> :in_progress
      _ -> :in_progress
    end
  end

  @badges %{
    not_started: {"●", :light_black, "not started"},
    in_progress: {"◐", :yellow, "in progress"},
    completed: {"✓", :green, "completed"},
    published: {"★", :cyan, "published"},
    locked: {"🔒", :red, "locked"}
  }

  @doc "An Owl-colored badge tag for a status."
  def badge(status) do
    {glyph, color, _label} = Map.fetch!(@badges, status)
    Owl.Data.tag(glyph, color)
  end

  @doc "Human label for a status."
  def label(status) do
    {_glyph, _color, label} = Map.fetch!(@badges, status)
    label
  end

  @doc "Legend line for the exercise list footer."
  def legend do
    [:not_started, :in_progress, :completed, :published, :locked]
    |> Enum.map(fn s -> [badge(s), " ", label(s)] end)
    |> Enum.intersperse("   ")
  end
end
